### PR TITLE
Add a "Hide duplicates" filter to the import preview grid

### PR DESCRIPTION
**Pull request by jss367**
Importing a card with 985 files where 984 are already in the library buries the one new photo in a wall of `DUPLICATE` tiles — you have to scroll the whole grid to find it.

## What changed

- New **"Hide duplicates (N)"** checkbox under the preview summary, **off by default**. Checking it filters the flagged files out of the preview grid.
- The control only appears once a preview has actually flagged duplicates (not in in-place/snapshot mode, not with "Skip duplicates" off) — otherwise there is nothing to hide and the checkbox would be a no-op.
- Toggling is purely a **view filter**: it re-renders from the last preview result, never re-running `/api/import/folder-preview` or `/api/import/check-duplicates`, and never changing what the import does.

## Transparency

Per `CORE_PHILOSOPHY.md`, the smaller grid must not read as "that is all there is":

- The summary line is untouched — still `985 files found · 984 already in your library (will be skipped) · 1 to copy`.
- The checkbox label carries the count being hidden: `Hide duplicates (984)`.
- Each folder header shows the visible count plus what it dropped: `104NCZ_8 (1) · 984 duplicates hidden`.
- If every file is a duplicate, the grid says `All 985 files are duplicates — uncheck "Hide duplicates" to see them.` instead of going blank.

## Screenshots

Off (default) → on:

`card-a (2)` with both tiles → `card-a (1) · 1 duplicate hidden` with only the new file.

## Tests

- New e2e case `test_import_preview_hide_duplicates_checkbox_filters_grid`: asserts the checkbox defaults to unchecked, that checking it removes the flagged file and shows the per-folder hidden count, that the summary still reports the full picture, and that unchecking restores the tile.
- Refactored the existing preview test onto a shared `TWO_FILE_PREVIEW_STUB` / `run_two_file_preview` helper rather than duplicating ~90 lines of fetch stubbing.
- `python -m pytest tests/e2e/test_import_page.py -q` → **31 passed**.
- `python -m pytest tests/test_workspaces.py vireo/tests/test_db.py vireo/tests/test_app.py vireo/tests/test_photos_api.py vireo/tests/test_edits_api.py vireo/tests/test_jobs_api.py vireo/tests/test_darktable_api.py vireo/tests/test_config.py -q` → **1949 passed, 1 failed**. The failure is `test_api_exiftool_status_reports_missing`, which fails identically on a clean tree at this commit (local ExifTool environment) and is unrelated to this change.

Verified in a real browser via Playwright screenshots of both states.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a “Hide duplicates” option to import previews.
  * Duplicate items can now be hidden from the preview grid without changing the overall preview summary.
  * Duplicate counts now appear in the filter label and folder group headers.
  * Shows an empty-state message when all items are hidden.
* **Bug Fixes**
  * The duplicate filter now resets appropriately across preview cycles (including no-duplicate outcomes and non-copy/special preview modes) to prevent stale selections.
* **Tests**
  * Added end-to-end coverage for toggling the duplicate filter, repreview behavior, and restoring the full preview.

### CODE REVIEW

File: vireo/templates/import.html [hunk 1]
```diff
@@ -1910,12 +1919,39 @@ <h3 id="folderBrowserTitle">Select Folder</h3>
...
+// flagged duplicates — before the duplicate check runs (or with dedup off)
+// there is nothing to hide, so the control stays hidden rather than
+// implying a filter that would do nothing.
+function updateHideDuplicatesControl(dupCount) {
+  const row = document.getElementById('hideDuplicatesRow');
+  const label = document.getElementById('hideDuplicatesLabel');
+  if (!row || !label) return;
+  row.style.display = dupCount > 0 ? '' : 'none';
+  label.textContent = 'Hide duplicates (' + dupCount.toLocaleString() + ')';
+}
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html:1947]: Good catch on the underlying issue — a stale checked value did survive the row being hidden, so a later duplicate-bearing card could filter itself without the user opting in. Fixed in 2e196d72.

I didn't take the proposed diff though, because resetting inside `updateHideDuplicatesControl` would regress a more common path. That function is called with `dupCount === 0` **twice on every preview**, before the answer is known:

- `clearImportPreviewGrid()` → `updateHideDuplicatesControl(0)`
- `renderImportPreviewGrid(files, [], null)` → the pre-check render, also 0

Only the third call carries the real count. And previews are auto-scheduled by `wireDestStructureInvalidation` on any import-option change — so with that diff, checking "Hide duplicates" and then nudging the folder template, destination, or verification mode would silently uncheck the box and flood every duplicate tile back into the grid.

Instead the opt-in is retired only where the duplicate picture is **final**:

- the check completed and returned zero duplicates
- "Skip duplicates" is off (nothing will be skipped)
- in-place / snapshot imports (no duplicate concept)

The fall-through path deliberately leaves it alone, since its count is still unknown there.

Three e2e tests: `test_hide_duplicates_resets_once_a_check_finds_no_duplicates` and `test_hide_duplicates_resets_when_dedup_is_turned_off` cover your scenario (both watched failing first), and `test_hide_duplicates_survives_a_repreview_that_still_has_duplicates` pins the behavior the suggested version would have broken.

File: vireo/templates/import.html [hunk 2]
```diff
@@ -1910,12 +1919,53 @@ <h3 id="folderBrowserTitle">Select Folder</h3>
     importThumbSchedulerCancel();
     importThumbSchedulerCancel = null;
   }
+  lastImportPreviewRender = null;
+  updateHideDuplicatesControl(0);
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html:1930]: Confirmed and fixed in 51dc0654 — thanks, this was a real gap the previous commit left behind.

Traced it: the `×` handler at `renderSources()` calls `clearImportPreviewGrid()` when `sources.length` hits 0, which hides the row but left `chkHideDuplicates.checked` set. Adding a different card then filtered it on arrival.

Retired the opt-in at that source-removal site specifically, not inside `clearImportPreviewGrid()` — as you noted, ordinary re-previews clear the grid too and must preserve the filter (that's what the earlier `test_hide_duplicates_survives_a_repreview_that_still_has_duplicates` pins).

Covered by `test_hide_duplicates_resets_when_the_last_source_is_removed`, watched failing first. The opt-in now retires in four places, all of them points where this card's duplicate picture is final or the card session has ended: completed zero-duplicate check, dedup off, in-place/snapshot import, and source list emptied.

File: vireo/templates/import.html [hunk 3]
```diff
@@ -1947,16 +2016,22 @@ <h3 id="folderBrowserTitle">Select Folder</h3>
   });
 
   Object.keys(groups).sort().forEach((subfolder) => {
+    const all = groups[subfolder];
+    const shown = hidingDuplicates ? all.filter((f) => !dupes.has(f.path)) : all;
+    if (!shown.length) return;
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html]: Confirmed — fixed in d39b87d9 (the repo's PR agent got to this one first, same `!files.length` placement I'd reached independently).

Rather than keep resolving these one at a time, I swept the rest of `previewImport` for the same class and found two more in 28992924: both pre-flight bail-outs — no sources, and no file extensions selected — hid the row while leaving `chkHideDuplicates` checked. The file-extension one is reachable in a few clicks: opt in, switch the preset to custom, clear the boxes, and the next card with duplicates arrives pre-filtered.

The opt-in now retires at six settled points, every place where this card's duplicate picture is final or its session has ended:

1. completed check returned zero duplicates
2. "Skip duplicates" off
3. in-place import
4. snapshot import
5. source list emptied
6. preview found no files
7. the two pre-flight bail-outs

The in-flight pre-check render still deliberately preserves it — that's the case `test_hide_duplicates_survives_a_repreview_that_still_has_duplicates` pins.

37 e2e tests, green twice in a row. One note: the new test restores the file-type preset on its way out, because a full-suite run showed it otherwise stranding a later test with an unpreviewable page.

[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html]: Correction to my previous reply: the count is **eight** call sites, not six — my list was already inconsistent with the number I put in front of it, and 28992924's commit message says "six" too. The code is right; the tally wasn't.

```
import.html:938   source list emptied
import.html:2332  bail-out: no sources
import.html:2340  bail-out: no file extensions selected
import.html:2384  preview completed with zero files
import.html:2394  snapshot import
import.html:2403  in-place import
import.html:2408  "Skip duplicates" off
import.html:2460  completed check returned zero duplicates
```

Eight settled conditions, one deliberate exclusion: the pre-check render inside an in-flight preview.

**Comment by jss367** [comment]
🤖 Routine fired on your latest `COMMENTED`-state review (the tally correction on the folder-headers thread), but the review body is empty and the wrapped comment is a self-clarification, not a request for changes.

Scanned the rest of the PR for anything unaddressed:
- CodeRabbit review on `2e196d7297` — the "reset checkbox when hiding the row" thread is resolved; you took a targeted `retireHideDuplicatesOptIn()` approach at eight settled call sites instead of the proposed in-place reset. Confirmed by CodeRabbit.
- Codex reviews on `2e196d7297` / `51dc06546e` / `d39b87d925` — the source-removal, empty-preview, and folder-header-visibility gaps were all fixed by `51dc0654`, `d39b87d9`, and `28992924` respectively.
- The one still-open CodeRabbit nitpick (adding an e2e edge case that asserts the "All X file(s) are duplicates" empty-state message when every file is filtered out) is marked "🔵 Trivial" — not fixing it unsolicited, but happy to add it if you'd like.

No commit pushed. Stand down.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vireo/templates/import.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html]: Fixed in 89133f68. A fully filtered folder now renders its header (`A (0) · 2 duplicates hidden`) and omits only the tiles, matching what partially-filtered folders already disclosed.

One consequence worth calling out: because headers always render now, the old `!grid.children.length` guard for the "all files are duplicates" notice became unreachable. That's handled — the notice now keys off whether any *tile* was rendered, and when every folder is fully filtered it replaces the wall of `(0) · N hidden` headers with the single message, which reads better than a column of zeroes.

File: vireo/templates/import.html [hunk 4]
```diff
@@ -1947,16 +2016,22 @@ <h3 id="folderBrowserTitle">Select Folder</h3>
   });
 
   Object.keys(groups).sort().forEach((subfolder) => {
+    const all = groups[subfolder];
+    const shown = hidingDuplicates ? all.filter((f) => !dupes.has(f.path)) : all;
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html]: Fixed in 9dc90446, and I verified it independently rather than relying on the accompanying test.

Confirmed your premise first: `/api/import/folder-preview` loops over each source with no cross-source dedup, and the UI's add-source guard only rejects an exact repeat (`!sources.includes(v)`), so `/card` + `/card/DCIM` really does yield the same path twice.

The fix replaces the path-set membership test with per-occurrence flags, walking `files` in reverse and consuming duplicate-count credits per path so it matches the server's "the later occurrence is the intra-import duplicate" ordering.

Driving it directly in the browser with your exact scenario — two occurrences of `1.jpg`, one flagged:

```
computeDuplicateFlags(files, ['/card/DCIM/1.jpg']) -> [false, true, false]
```

Exactly one occurrence flagged, and it's the later one. With the filter on, 2 tiles survive (the to-copy `1.jpg` plus `2.jpg`) and the grid no longer claims everything is a duplicate while the summary reports one to copy.

**Comment by jss367** [comment]
🤖 Routine fired on another empty-body `COMMENTED`-state review from you. Both of the two most recent ones wrap your own replies to Codex threads on commit `89133f68`, announcing fixes you'd already pushed — not requests for changes:

- `4793794880` (04:46:04) → your reply to the fully-filtered folder-headers thread ("Fixed in 89133f68…")
- `4793795045` (04:46:07) → your reply to the overlapping-sources thread ("Fixed in 9dc90446…")

Nothing changed vs. the earlier stand-down: the one still-open CodeRabbit item on this PR is the trivial e2e nitpick, and I'm not touching it unsolicited.

No commit pushed. Stand down.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jss367** [comment]
🤖 Routine fired again on empty-body `COMMENTED`-state reviews (both from you, 04:46:04 and 04:46:07). They wrap the two reply comments on the fully-filtered-folder-headers thread (fixed in `89133f68`) and the overlapping-sources thread (fixed in `9dc90446`) — self-clarifications, not requests for changes.

State of the PR at head `89133f68`:
- All Codex/CodeRabbit threads that flagged real bugs (checkbox reset, source-list emptied, empty preview, pre-flight bail-outs, fully-filtered folder headers, overlapping sources) are addressed by commits already on the branch.
- The one remaining CodeRabbit item is the 🔵 Trivial nitpick asking for an e2e edge case that asserts the "All X file(s) are duplicates" empty-state message when every file is filtered out — still not fixing unsolicited.
- CI: all completed checks green, `test-os` (ubuntu/mac/windows) still in progress on this head.

No commit pushed. Stand down.

---
_Generated by [Claude Code](https://claude.ai/code)_
